### PR TITLE
EBS-1070 Fix #427 VSDK fix required to handle empty system password during Lua to VSDK upgrade

### LINKED ISSUE #427: "[March 05, 2022] [BUG] retrieve_credentials() returns incorrect KeyPairCredentials object if credential_supplier has empty password"
** Expected Behavior**
Even if `credentials_supplier` dict  has empty password is should return PasswordCredentials() object but instead KeyPairCredentials() is returned.

** Actual Behavior**
For EBS plugin implemented in vsdk, we use passwordCredentialsSupplier for System Password field which is retrieved from UI and used in our workflows. But this field is optional field, where user may not provide it in most of the cases. So when we upgrade the plugin from Lua to VSDK and empty password is passed in for System password field. After upgrade, when VSDK code retrieves it using retrieve_credentials() fails with AttributeError as retrieve_credentials() returns the KeyPairCredentials() instead of PasswordCredentials(). This is because there is a check `credentials_result.password != ""` which is incorrect, instead it should be `"password" in credentials_supplier`

**Steps To Reproduce the Problem**
Call the retrieve_credentials() with dict {'password': '', 'type': 'NamedPasswordCredential'},
```python
from dlpx.virtualization import libs
vault = {'password': '', 'type': 'NamedPasswordCredential'}
password = libs.retrieve_credentials(vault).password # We hit AttributeError as it returns KeyPairCredentials object which doesn't have attribute password.
```

1. Use EBS Lua plugin to link and provision datasets with System Password field empty
2. Disable datasets.
3. Upgrade Lua to VSDK plugin
4. Refresh the environments
5. Enable datasets after successful upgrade.
6. We hit AttributeError when dataset starts to run on vsdk plugin due to this issue

### PR DESCRIPTION
**Pull request by Balamuruhan**
# Problem:

For EBS plugin implemented in vsdk, we use passwordCredentialsSupplier for System Password field which is retrieved from UI and used in our workflows. But this field is optional field, where user may not provide it in most of the cases. So when we upgrade the plugin from Lua to VSDK and empty password is passed in for System password field. After upgrade, when VSDK code retrieves it using retrieve_credentials() it fails with AttributeError as retrieve_credentials() returns the KeyPairCredentials() instead of PasswordCredentials(). 

Call the retrieve_credentials() with dict {'password': '', 'type': 'NamedPasswordCredential'},
```python
from dlpx.virtualization import libs
vault = {'password': '', 'type': 'NamedPasswordCredential'}
password = libs.retrieve_credentials(vault).password # We hit AttributeError as it returns KeyPairCredentials object which doesn't have attribute password.
```
# Solution:

This is because there is a check credentials_result.password != "" which is incorrect as password in this failure is observed when an empty string used from plugin, instead we tried to check whether there is a attribute "password" in the `credentials_result` object using the condition,
```python
if hasattr(credentials_result, "password")
```
But irrespective of whether it is `PasswordCredentials` or `KeyPairCredentials` the password attribute will be present even when it's a key pair result,

This is the protobuf definition of that result object:
```
message CredentialsResult {
    message KeyPair {
        string private_key = 1;
        string public_key = 2;
    }
    string username = 1;
    oneof result {
        string password = 2;
        KeyPair key_pair = 3;
    }
}
```
So we currently check whether `key_pair.private_key` and `key_pair.public_key` are empty to consider `credentials_result` as PasswordCredentials otherwise it can be considered as KeyPairCredentials.

Thanks to @rasantel for the help and suggestion.

# Testing

**EBS upgrade:**
Tested EBS plugin with Datasets by performing upgrade from Lua to VSDK with empty system password with this change and Datasets are recovered successfully after upgrade.

**Blackbox runs:**
appdata_python_samples
QARunner -d dlpxdc.co group -g dlpx-master nightly -r appdata_python_samples remote --run-params='-p virt-sdk-repo=https://github.com/Balamuruhan/virtualization-sdk.git -p virt-sdk-branch=develop -p qa-plugin-repo=https://gitlab.delphix.com/balamuruhan.suriyakumar/qa-appdata-toolkits.git'

http://selfservice.jenkins.delphix.com/job/blackbox-self-service/15404/consoleFull

APPDATA_PYTHON_DIRECT_CENTOS73
QARunner -d dlpxdc.co group -g dlpx-master nightly -r appdata_basic --run-config APPDATA_PYTHON_DIRECT_CENTOS73 remote --run-params='-p virt-sdk-repo=https://github.com/Balamuruhan/virtualization-sdk.git -p virt-sdk-branch=develop -p qa-plugin-repo=https://gitlab.delphix.com/balamuruhan.suriyakumar/qa-appdata-toolkits.git'

http://selfservice.jenkins.delphix.com/job/blackbox-self-service/15401/consoleFull

APPDATA_PYTHON_STAGED_CENTOS73
QARunner -d dlpxdc.co group -g dlpx-master nightly -r appdata_basic --run-config APPDATA_PYTHON_STAGED_CENTOS73 remote --run-params='-p virt-sdk-repo=https://github.com/Balamuruhan/virtualization-sdk.git -p virt-sdk-branch=develop -p qa-plugin-repo=https://gitlab.delphix.com/balamuruhan.suriyakumar/qa-appdata-toolkits.git'

http://selfservice.jenkins.delphix.com/job/blackbox-self-service/15405/consoleFull

**For unit tests:**
Failure is due to the windows-latest image that doesn't support Python 2.7, I have submitted TOOLS ticket for the new image to be used.
https://delphix.atlassian.net/browse/TOOL-13254

### CODE REVIEW

**Review by mrburke — COMMENTED** [review]
I think this looks good but can take another pass when the unit tests are updated as well

**Review by rasantel — APPROVED** [review]

File: libs/src/main/python/dlpx/virtualization/libs/libs.py [hunk 1]
```diff
@@ -468,7 +468,10 @@ def retrieve_credentials(credentials_supplier):
     response = internal_libs.retrieve_credentials(credentials_request)
     response_to_str(response)
     credentials_result = _handle_response(response)
-    if credentials_result.password != "":
+    if (
+        credentials_result.key_pair.private_key == "" and
+        credentials_result.key_pair.public_key == ""
```
Comment (nhlien93) [inline libs/src/main/python/dlpx/virtualization/libs/libs.py:473]: `if not credentials_result.key_pair.private_key and not credentials_result.key_pair.public_key:`?
Don't know if this actually works but ideally we shouldn't be comparing empty strings.

Comment (Balamuruhan) [reply libs/src/main/python/dlpx/virtualization/libs/libs.py]: I have made the changes as suggested

**Comment by vimleshmishra** [comment]
Able to reproduce this issue in-house, the suggested solution will definitely work.

But one alternate decent solution could be simply checking below before returning "PasswordCredentials"

`if hasattr(credentials_result, "password")`

The actual bug is the engine was returning empty password **[but not key_pair]** while the plugin is returning "PasswordCredentials" only if non-empty password OR else returns "KeyPairCredentials". While in this case since the engine returns empty password plugin will return the "KeyPairCredentials" which won't have the "password" attribute.

File: libs/src/main/python/dlpx/virtualization/libs/libs.py [hunk 2]
```diff
@@ -468,7 +468,10 @@ def retrieve_credentials(credentials_supplier):
     response = internal_libs.retrieve_credentials(credentials_request)
     response_to_str(response)
     credentials_result = _handle_response(response)
-    if credentials_result.password != "":
+    if (
```
Comment (ravi-cm) [inline libs/src/main/python/dlpx/virtualization/libs/libs.py:471]: nit: can you please add a comment to say why this check is needed and that way anyone modifying this part of code will know what to do/not do?

**Comment by ravi-cm** [comment]
> Able to reproduce this issue in-house, the suggested solution will definitely work.
> 
> But one alternate decent solution could be simply checking below before returning "PasswordCredentials"
> 
> `if hasattr(credentials_result, "password")`
> 
> The actual bug is the engine was returning empty password **[but not key_pair]** while the plugin is returning "PasswordCredentials" only if non-empty password OR else returns "KeyPairCredentials". While in this case since the engine returns empty password plugin will return the "KeyPairCredentials" which won't have the "password" attribute.

Would this work if, for some reason a KeyPairCredentials with a password field is returned since protobuf definition allows that?

**Review by ravi-cm — APPROVED** [review]
lgtm.

File: libs/src/main/python/dlpx/virtualization/libs/libs.py (hunk 2 above)
Comment (Balamuruhan) [reply libs/src/main/python/dlpx/virtualization/libs/libs.py:476]: Thank you Ravi. Sure, I've added the comment that explains about this change.

**Comment by vimleshmishra** [comment]
> Would this work if, for some reason, a KeyPairCredentials with a password field is returned since protobuf definition allows that?

But protobuf also allow password along with KeyPairCredentials private and public keys.
While the engine is used to make sure that only one should be set either password or private/public key

File: libs/src/main/python/dlpx/virtualization/libs/libs.py [hunk 3]
```diff
@@ -468,7 +468,15 @@ def retrieve_credentials(credentials_supplier):
     response = internal_libs.retrieve_credentials(credentials_request)
     response_to_str(response)
     credentials_result = _handle_response(response)
-    if credentials_result.password != "":
+    # As protobuf definition of credentials_result object has all the
+    # attributes private_key, public_key and password irrespective of
+    # whether it is a keypair or a password credential type, we consider the
+    # object to be password credential type if private_key and public_key is
+    # not set.
```
Comment (nhlien93) [inline libs/src/main/python/dlpx/virtualization/libs/libs.py:475]: for multiline comments add a `#` line before and after. so:
```
    #
    # As protobuf definition of credentials_result object has all the
    # attributes private_key, public_key and password irrespective of
    # whether it is a keypair or a password credential type, we consider the
    # object to be password credential type if private_key and public_key is
    # not set.
    #
```

Comment (Balamuruhan) [reply libs/src/main/python/dlpx/virtualization/libs/libs.py:476]: I've made changes as suggested.

**Pull request merged**